engines/syguspdr.cpp

```diff
@@ -0,0 +1,1542 @@
+{
+  assert(solver_context_ == 0);
+  assert(i < frame_labels_.size());
+  assert(constraint.disjunction);
+  if (has_assumptions || options_.sygus_use_operator_abstraction_ == 1) {
+    // options_.sygus_use_operator_abstraction_ == 2 will not introduce new inputs
+    // that must be kept
+    assert(ts_.no_next(constraint.term));
+  } else
+    assert(ts_.only_curr(constraint.term));
+
+  if (new_constraint) {
+    for (size_t j = 1; j <= i; ++j) {
+      std::vector<IC3Formula> & Fj = frames_.at(j);
+      size_t k = 0;
+      for (size_t l = 0; l < Fj.size(); ++l) {
+        if (!subsumes(constraint, Fj[l])) {
+          Fj[k++] = Fj[l];
+        }
+      }
+      Fj.resize(k);
+    }
+  }
+
+  assert(i > 0);  // there's a special case for frame 0
+
+  constrain_frame_label(i, constraint);
+  frames_.at(i).push_back(constraint);
+}
+
+// -----------------------------------------------------------------
+// Require the capability to reset solver
+// -----------------------------------------------------------------
+
+void SygusPdr::reset_solver()
+{
+  assert(solver_context_ == 0);
+
+  // for syguspdr, we must be able to reset solver
+  assert (!failed_to_reset_solver_);
+
+  try {
+    solver_->reset_assertions();
+
+    // Now need to add back in constraints at context level 0
+    logger.log(2, "SygusPdr: Reset solver and now re-adding constraints.");
+
+    // define init, trans, and bad labels
+    assert(init_label_ == frame_labels_.at(0));
+    solver_->assert_formula(
+        solver_->make_term(Implies, init_label_, ts_.init()));
+
+    solver_->assert_formula(
+        solver_->make_term(Implies, trans_label_, ts_.trans()));
+
+    // assume property in pre-state
+
+    solver_->assert_formula(solver_->make_term(Implies, bad_label_, bad_));
+
+    for (size_t i = 0; i < frames_.size(); ++i) {
+      for (const auto & constraint : frames_.at(i)) {
+        constrain_frame_label(i, constraint);
+      }
+    }
+  }
+  catch (SmtException & e) {
+    logger.log(1,
+               "Failed to reset solver (underlying solver must not support "
+               "it). Disabling solver resets for rest of run.");
+    throw e; // SyGuS requires reset solver capability 
+  }
+
+  num_check_sat_since_reset_ = 0;
+}
+
+// difference : I don't need to conjuct with property, 
+// because it is already added as a lemma/constraint
+Term SygusPdr::get_frame_term(size_t i) const
+{
+  // TODO: decide if frames should hold IC3Formulas or terms
+  //       need to special case initial state if using IC3Formulas
+  if (i == 0) {
+    // F[0] is always the initial states constraint
+    return ts_.init();
+  }
+
+  Term res = solver_true_;
+  for (size_t j = i; j < frames_.size(); ++j) {
+    for (const auto &u : frames_[j]) {
+      res = solver_->make_term(And, res, u.term);
+    }
+  }
+  return res;
+}
+
+// Difference : I need it to call the new constrain_frame
+// which allow input in the constraints
```

Comment:
This can be fixed with the pass that converts to state variables. I'm just commenting here as a reminder to remove that after the pass is implemented.